cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -37,13 +39,41 @@
 namespace arrow {
 namespace compute {
 
-// Remove this when multiple keys and/or types is supported
-typedef int32_t KeyType;
+template <typename T, typename V = typename T::value_type>
+inline typename T::const_iterator std_find(const T& container, const V& val) {
+  return std::find(container.begin(), container.end(), val);
+}
+
+template <typename T, typename V = typename T::value_type>
+inline bool std_has(const T& container, const V& val) {
+  return container.end() != std_find(container, val);
+}
+
+typedef uint64_t KeyType;
+typedef uint64_t TimeType;
+typedef uint64_t HashType;
 
 // Maximum number of tables that can be joined
 #define MAX_JOIN_TABLES 64
 typedef uint64_t row_index_t;
 typedef int col_index_t;
+typedef std::vector<col_index_t> vec_col_index_t;
+
+template <typename T, enable_if_t<std::is_integral<T>::value, bool> = true>
+static inline uint64_t norm_value(T t) {
```

Comment:
Can you add some doc string for what this function is doing?

Reply:
It normalizes the value to `uint64_t` while preserving order. I'll add this in a doc string.